packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -355,6 +358,39 @@ class AndroidDevice extends Device {
   }
 
   @override
+  Future<bool> restartApp(
+    ApplicationPackage package,
+    LaunchResult result, {
+    String mainPath,
+    Observatory observatory
+  }) async {
+    Directory tempDir = await Directory.systemTemp.createTemp('flutter_tools');
```

Comment:
This is the pattern that the current `flutter refresh` command uses. We may want to put more things into `build/`, and use a temp dir less often. I'd still want most of the build artifacts in `build/` to be semantic wrt to the user - something that they would understand and identify as useful to their project.